### PR TITLE
Mui-datatables PoC

### PR DESCRIPTION
**Pull request by quorth0n**
Fully functioning:
- [x] A data table adhering to our stylesheets
- [x] Details pain (sidebar)
- [x] Show/hide columns
- [x] Dumb stuff I don't think we'll need like printing and CSV export
- [x] Server-side pagination
- [x] Sever-side sorting

PoC ready (implementation is there with a placeholder/botched func, should be trivial to implement):
- [x] Custom column rendering functions, such as for `location`

Possible but no PoC yet:
~~- [ ] Server-side filtering (search bar or dropdown[²] matching )~~

#### ² This plugin requires mui >3.2.0. This project is currently on 1.1, which is getting a bit old. It seems most features still work, but multiselect filtering, which I think would be the most fitting, does not. I would assume we would run into other issues down the road using an older version, so it might be a good idea to update?
[²]:#-multiselect

### CODE REVIEW

**Comment by quorth0n** [comment]
image

**Comment by quorth0n** [comment]
Complete list of customizable options for the plugin: https://github.com/gregnb/mui-datatables#options. Forgot to mention, action drawer would probably be done through `customToolbar`/`Select`

**Comment by ZavenArra** [comment]
@whirish have you checked out all the server side filtering already supported by LoopBack in our current admin api in this repo?

**Comment by ZavenArra** [comment]
@whirish Can you resolve the conflicts in this PR now that the image scrubber is merged ?

**Comment by quorth0n** [comment]
@deepwinter thanks for the review! Yeah, I saw that lb already has endpoints set up to handle the serverside operations for us, which is awesome. I didn't want to implement them yet in case there was some other reason we couldn't use this lib, but if it looks okay I'll get started on it now!

**Comment by quorth0n** [comment]
In the latest commit I have server-side pagination almost working. The main issue is here https://github.com/Greenstand/treetracker-admin/pull/45/files#diff-8914dfd42807bb668c5203b335fff52fR107. Retrieving the count through `this.state.treeCount` or `this.props.treeCount` both return the count after rowsPerPage, so, if rowsPerPage is 25, 25 is returned for the total treeCount, and the pagination buttons are disabled. If this behavior hasn't been properly implemented yet I can look into it, but I think it's working on master.
Past that, I couldn't get managing the `page` state to work through rematch (it was resulting in something similar to the bug described above), so I managed it locally, which works fine. Not sure if this is a good practice with this library.

**Comment by quorth0n** [comment]
@thiagodemellobueno do you think you could provide some insight on the above post when you get a chance? Thanks :slightly_smiling_face:

**Comment by thiagodemellobueno** [comment]
I too have noticed this, but all I know is it once worked.

**Comment by ZavenArra** [comment]
@whirish I am not sure what the question is.. rowsPerPage is the number of rows in the page, and treeCount is the number of trees on the page.  So they will generally be the same number as you paging through.

Pagination works just fine on the master branch, except that the 'from-to' text doesn't seem to have the 'to' value calculated.

If you are wanting the total number of trees in the database, that would have to queried separately.   Pagination works by selected with a start index and limit via the Loopback API each time < or > is pressed.  The admin app won't know the total number of trees in the database without a separate query.

**Comment by quorth0n** [comment]
@deepwinter @thiagodemellobueno ready for a review I think

**Comment by ZavenArra** [comment]
@whirish Sorry I broke your PR.

**Comment by quorth0n** [comment]
@deepwinter fixed conflicts

**Comment by quorth0n** [comment]
@deepwinter  @thiagodemellobueno any updates on this?

File: client/package-lock.json [hunk 1]
```diff
@@ -9787,6 +9835,21 @@
       "resolved": "https://registry.npmjs.org/ms/-/ms-2.1.1.tgz",
       "integrity": "sha512-tgp+dl5cGk28utYktBsrFqA7HKgrhgPsg6Z/EfhWI4gl1Hwq8B/GmY/0oXZ6nF8hDVesS/FpnYaD/kOWhYQvyg=="
     },
+    "mui-datatables": {
```
Comment (mckornfield) [inline client/package-lock.json:9751]: Probably need to rebase and rerun `npm install` with this component

File: client/src/components/TreeTable/TreeTable.js [hunk 2]
```diff
@@ -40,148 +32,122 @@ const styles = theme => ({
...
+    const columnData = [
+      { name: 'id', label: 'Id', options: { sortable: true } },
+      { name: 'timeCreated', label: 'Creation', options: { sortable: true } },
+      { name: 'timeUpdated', label: 'Updated', options: { sortable: true } },
+      {
+        label: 'Location',
+        options: {
+          sortable: false,
+          customBodyRender: () => {
+            return <span>botched rendering func</span>; // TODO: replace
```
Comment (mckornfield) [inline client/src/components/TreeTable/TreeTable.js]: is this todo for this PR or a future one?

File: client/src/components/TreeTable/TreeTable.js [hunk 3]
```diff
@@ -40,148 +32,122 @@ const styles = theme => ({
...
+        const id = treesArray[rowMeta.dataIndex].id;
+        this.toggleDrawer(id);
+      },
+      serverSide: true,
+      count: this.props.treeCount,
+      page: this.state.page,
+      rowsPerPage: this.props.rowsPerPage,
+      rowsPerPageOptions: [25, 50, 75, 100, 250, 500],
+      onTableChange: (action, tableState) => {
+        console.log(action, tableState); // TODO: remove
```
Comment (mckornfield) [inline client/src/components/TreeTable/TreeTable.js]: Probably should remove this

File: client/src/components/TreeTable/TreeTable.js [hunk 4]
```diff
@@ -40,148 +32,122 @@ const styles = theme => ({
...
-      )
-    }
-    this.setState({ selected: newSelected })
-  }
-
-  toggleDrawer = (event, id) => {
-    event.preventDefault();
-    if (id !== undefined) this.props.getTreeAsync(id);
+  toggleDrawer = id => {
+    console.log(id);
```
Comment (mckornfield) [inline client/src/components/TreeTable/TreeTable.js]: Needed?

File: server/datasources.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (mckornfield) [inline server/datasources.json]: Please remove this. Also add it to the git ignore if it's not already ignored

File: package-lock.json [hunk 6]
```diff
@@ -0,0 +1,4251 @@
+{
```
Comment (mckornfield) [inline package-lock.json]: Should this be in the root, or should it have changed in the `client` folder?

File: client/src/models/trees.js [hunk 7]
```diff
@@ -9,118 +8,147 @@ const trees = {
...
     async sortTrees(payload, rootState) {
       const { page, rowsPerPage, order } = rootState.trees;
-      const newOrder = (order === 'asc') ? 'desc' : 'asc';
-      const query = `${API_ROOT}/trees?filter[order]=${payload.orderBy} ${newOrder}&filter[limit]=${rowsPerPage}&filter[skip]=${page * rowsPerPage}&filter[fields][lat]=true&filter[fields][lon]=true&filter[fields][id]=true&filter[fields][timeCreated]=true&filter[fields][timeUpdated]=true`;
-      Axios.get(query)
-        .then((response) => {
-          this.getTrees(response.data, { page: page, rowsPerPage: rowsPerPage, orderBy: payload.orderBy, order: newOrder });
+      const newOrder = order === 'asc' ? 'desc' : 'asc';
+      const query = `${API_ROOT}/trees?filter[order]=${
+        payload.orderBy
```
Comment (mckornfield) [inline client/src/models/trees.js:139]: Formatting is kinda strange... I think you should keep the interpolated values on the same line, and keep the width about the same, but this is more of a nit

File: client/src/models/trees.js [hunk 8]
```diff
@@ -9,118 +8,147 @@ const trees = {
...
+      if (
+        (rootState.trees.byId[payload.id] &&
+          rootState.trees.byId[payload.id].location &&
+          rootState.trees.byId[payload.id].location.lat !== payload.lat &&
+          rootState.trees.byId[payload.id].location.lon !== payload.lon) ||
+        (!rootState.trees.byId[payload.id] ||
+          !rootState.trees.byId[payload.id].location)
+      ) {
+        const query = `https://nominatim.openstreetmap.org/reverse?format=jsonv2&lat=${
+          payload.latitude
```
Comment (mckornfield) [inline client/src/models/trees.js:118]: Same comment here as below, string interpolation seems strange

File: client/src/models/trees.js [hunk 9]
```diff
@@ -9,118 +8,147 @@ const trees = {
...
-        .then((response) => {
-          this.getTrees(response.data, { page: page, rowsPerPage: rowsPerPage, orderBy: orderBy, order: order });
-        })
+    async getTreesWithImagesAsync({
+      page,
+      rowsPerPage,
+      orderBy = 'id',
+      order = 'asc'
+    }) {
+      const query = `${API_ROOT}/trees?filter[order]=${orderBy} ${order}&filter[limit]=${rowsPerPage}&filter[skip]=${page *
```
Comment (mckornfield) [inline client/src/models/trees.js:70]: A method to make this query string might be good (I realize these are mostly formatting changes though on your part)

**Comment by thiagodemellobueno** [comment]
@whirish - I have a launch this week thats keepin my headspace busy, but maybe have time friday to check it out.

**Review by mckornfield — COMMENTED** [review]
Overall functionally looks good, some comments around styling (I see you added semicolons and changed the formatting in some places, which I don't know if that was necessary)...

Also make sure to remove the datasources file you committed. Also I don't know if the package*.json files are in the same place as they once were

File: client/src/components/TreeTable/TreeTable.js (hunk 2 above)
Comment (quorth0n) [reply client/src/components/TreeTable/TreeTable.js]: Future, location rendering is not implemented on master to my knowledge

File: client/src/components/TreeTable/TreeTable.js (hunk 4 above)
Comment (quorth0n) [reply client/src/components/TreeTable/TreeTable.js]: I think it was leftover from the original toggleDrawer function :) but sure

File: package-lock.json (hunk 6 above)
Comment (quorth0n) [reply package-lock.json]: It should be fine here, it changed because of adding the mui-datatables pkg

**Comment by quorth0n** [comment]
All formatting changes were done automatically by my VSCode config, I submitted my last PR with similar changes too. Personally, I don't have any preference as far as formatting goes but I would rather have a consistent style being enforced. Is there any linting or formatting guides set up for this project? An eslint/prettier config would be nice.

**Comment by quorth0n** [comment]
@mckornfield Thanks for the review :) I left a few responses and will implement your requests

**Comment by mckornfield** [comment]
No problem, feel free to disregard some if they were formatting changes (I realized partway through)

**Comment by mckornfield** [comment]
I have no idea what the formatting setup is for the project either 😆 but I agree a checked in, standard format would be good

**Comment by thiagodemellobueno** [comment]
We had an issue for establishing formatting and linting in the old repo before it was migrated.
Note that this is the first month that this code is updated by soomeone other than me or jparaujo in the last year and a half :)

I spun up this issue https://github.com/Greenstand/treetracker-admin/issues/48 
and made some reccos.
Having an `.editorconfig` in there might also be quite helpful.

**Comment by ZavenArra** [comment]
Hey guys, I was looking to review and merge this PR for possible inclusion in our upcoming release, but it doesn't seem to be running correctly.  Please advise
image

**Comment by quorth0n** [comment]
@deepwinter Sorry about the delay, the school year is wrapping up for me and I've been preparing for finals. I will try to take a look this weekend but if anyone gets to it before me that's fine too.

**Comment by quorth0n** [comment]
@deepwinter merged with latest master and I believe your issue should be fixed. Also, server side sorting is now working. Please test & give any feedback at your leisure.

Btw, nice work with the image scrubber. Also, I saw some recent commits for formatting changes, is there a formatting guide in place now? Root level eslint still ignores `client/` :thinking:

**Pull request merged**